avalon/tvpaint/pipeline.py

```diff
@@ -28,8 +36,145 @@ def uninstall():
 
     """
 
-    pyblish.deregister_host("tvpaint")
+    pyblish.api.deregister_host("tvpaint")
+
+
+def containerise(name, namespace, nodes, context, loader):
+    data = {
+        "schema": "avalon-core:container-2.0",
+        "name": name,
+        "namespace": namespace,
+        "loader": str(loader),
+        "representation": str(context["representation"]["_id"])
+    }
+    return json.dumps(data)
+
+
+@contextlib.contextmanager
+def maintained_selection():
+    # TODO implement logic
+    try:
+        yield
+    finally:
+        pass
+
+
+def workfile_metadata(metadata_key):
+    """Read metadata for specific key from current project workfile.
+
+    Pipeline use function to store loaded and created instances within keys
+    stored in `SECTION_NAME_INSTANCES` and `SECTION_NAME_CONTAINERS`
+    constants.
+
+    Args:
+        metadata_key (str): Key defying which key should read. It is expected
+            value contain json serializable string.
+    """
+    george_script = (
+        "tv_readprojectstring \"{}\" \"{}\" \"[]\""
+    ).format(METADATA_SECTION, metadata_key)
+    json_string = lib.execute_george(george_script)
+    if json_string:
+        data = json.loads(json_string)
+    else:
+        data = []
+    return data
+
+
+def write_workfile_metadata(metadata_key, value):
+    """Write metadata for specific key into current project workfile.
+
+    George script has specific way how to work with quotes which should be
+    solved automatically with this function.
+
+    Args:
+        metadata_key (str): Key defying under which key value will be stored.
+        value (dict,list,str): Data to store they must be json serializable.
+    """
+    if isinstance(value, (dict, list)):
+        value = json.dumps(value)
+
+    if not value:
+        value = ""
+
+    # Handle quotes in dumped json string
+    value = (
+        value
+        # Replace both quotes with placeholder first
+        .replace("'", "{single_quote}").replace("\"", "{quote}")
+        # Replace plaholders with right values for george script
+        .replace("{single_quote}", "'\"'").replace("{quote}", "\"'\"")
+    )
+
+    george_script = (
+        "tv_writeprojectstring \"{}\" \"{}\" \"{}\""
+    ).format(METADATA_SECTION, metadata_key, value)
+    return lib.execute_george_through_file(george_script)
+
+
+def list_instances():
+    return workfile_metadata(SECTION_NAME_INSTANCES)
+
+
+def _write_instances(data):
+    return write_workfile_metadata(SECTION_NAME_INSTANCES, data)
 
 
 def ls():
-    return []
+    return workfile_metadata(SECTION_NAME_CONTAINERS)
+
+
+class Creator(api.Creator):
+    def __init__(self, *args, **kwargs):
+        super(Creator, self).__init__(*args, **kwargs)
+        # Replace unimportant `id` value "pyblish.avalon.instance" with
```

Comment:
I'd prefer if we keep this consistent with other hosts, of course we need uuid, but add "id" as we do for other hosts as well.

Reply:
It's ok for me just didn't see reason why to add more data to each item `"pyblish.avalon.container"` when is never used.